### DISCUSSION TITLE
Automatic regression tests

### DISCUSSION THREAD

**Discussion by dhschall** [gem5/gem5#446] [gem5-dev]
After today's developers meeting discussion, I was wondering if there is a simple way to add some sort of regression test to the existing testing infrastructure. Those are orthogonal to the normal unit tests and will report some common statistics like IPC, Branch MPKI, I-cache or D-cache MPKI, and simulated instructions per second (KIPS/MIPS).

In addition, the test can issue a warning if one of the statistics exceeds a certain range (for simulation time in particular).

I only thought about a few benchmarks for the most common ISA's and CPU cores. (I.e. x86, Arm, RISC-V, O3, atomic, timing)
 
This would have three main advantages.
1. For developers to quickly encounter performance degradations. Either simulated performance or simulation speed
2. Would help reviewers to quickly see if the change alters behavior or degrades performance.
3. For people new to gem5 it would give them a reference to see spot misconfigurations. (There are too many parameters you can mess with)

Also, it would be great to see those metrics evolve over time (in addition to stars and clones). It might help to speed up the simulation time of gem5 and improve the models.

**Comment by BobbyRBruce** [comment]
I agree that testing such as this would be a great benefic. Here at UC Davis we've discussed implementing stuff like this to ensure contributed code isn't making things work in a non-functional way (slowing it down or whatever else).

Another reason is we were keen on the idea it'd allow us to give better guarantees on the fidelity of a component we distribute. We can make tests for components to ensure their simulated statistics are within a certain bounds of their real-world equivalent. This could automatically be verified by developers, as well as by users, if they wished.

I think there's always been some enthusiasm to do something like this, but no one who can do it has found the time. I think if you wanted to do it, it'd be considered a good contribution.

**Comment by BobbyRBruce** [comment]
I've created issue #702 from this. I don't believe there is much more to discuss. Regardless as to how they are implemented, such tests would be of use. I'll close this discussion. Anyone is free to pick up the issue and work on it.
